[SPARK-23447][SQL] Cleanup codegen template for Literal

rednaxelafx opened the pull request:
## What changes were proposed in this pull request?

Cleaned up the codegen templates for `Literal`s, to make sure that the `ExprCode` returned from `Literal.doGenCode()` has:
1. an empty `code` field;
2. an `isNull` field of either literal `true` or `false`;
3. a `value` field that is just a simple literal/constant.

Before this PR, there are a couple of paths that would return a non-trivial `code` and all of them are actually unnecessary. The `NaN` and `Infinity` constants for `double` and `float` can be accessed through constants directly available so there's no need to add a reference for them.

Also took the opportunity to add a new util method for ease of creating `ExprCode` for inline-able non-null values.

## How was this patch tested?

Existing tests.

Commit:
Cleanup codegen templates for Literals: make sure the `code` field is empty and the `value` field is a simple literal.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution/930/
Test PASSed.

SparkQA commented:
**[Test build #87503 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/87503/testReport)** for PR 20626 at commit [`68edf0f`](https://github.com/apache/spark/commit/68edf0f3463daed3bb7042becb333788b22b23b0).

SparkQA commented:
**[Test build #87503 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/87503/testReport)** for PR 20626 at commit [`68edf0f`](https://github.com/apache/spark/commit/68edf0f3463daed3bb7042becb333788b22b23b0).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/87503/
Test FAILed.

rednaxelafx commented:
jenkins retest this please

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution/933/
Test PASSed.

rednaxelafx commented:
So I was able to find quite a few cases where the `DUMMY` placeholder caught uses of the `value` field outside of appropriate null-checked regions. I'll check the individual cases and then update this PR.

SparkQA commented:
**[Test build #87505 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/87505/testReport)** for PR 20626 at commit [`68edf0f`](https://github.com/apache/spark/commit/68edf0f3463daed3bb7042becb333788b22b23b0).

Commit:
Replace DUMMY with an actual default value (albeit unused at runtime) to allow compilation of the pattern:
  isNull ? defaultValue : value
where "value" has to be something that can be resolved.

rednaxelafx commented:
cc @cloud-fan @hvanhovell 
Note: this is for master and branch-2.3 post 2.3.0 release.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution/934/
Test PASSed.

SparkQA commented:
**[Test build #87506 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/87506/testReport)** for PR 20626 at commit [`d709e24`](https://github.com/apache/spark/commit/d709e246d99c0d821238afda1b203b9880eb1ed1).

rednaxelafx commented:
Ah...I see, there are more places where they're statically referencing some variable but dynamically those variables would always be null. I'll update the PR later to fix those places as well.

SparkQA commented:
**[Test build #87505 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/87505/testReport)** for PR 20626 at commit [`68edf0f`](https://github.com/apache/spark/commit/68edf0f3463daed3bb7042becb333788b22b23b0).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/87505/
Test FAILed.

SparkQA commented:
**[Test build #87506 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/87506/testReport)** for PR 20626 at commit [`d709e24`](https://github.com/apache/spark/commit/d709e246d99c0d821238afda1b203b9880eb1ed1).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/87506/
Test FAILed.

hvanhovell commented:
You are going to need to 'type' null values for this work, I think casting would be enough.

rednaxelafx commented:
@hvanhovell Thanks a lot! You're absolutely right. Update the PR accordingly. It's passing tests in my local testing and hopefully it'll pass the Jenkins tests as well.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution/938/
Test PASSed.

SparkQA commented:
**[Test build #87512 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/87512/testReport)** for PR 20626 at commit [`10419de`](https://github.com/apache/spark/commit/10419dec3f55e16075c4a9f10df837a35225b05e).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution/939/
Test PASSed.

SparkQA commented:
**[Test build #87513 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/87513/testReport)** for PR 20626 at commit [`2d89ea4`](https://github.com/apache/spark/commit/2d89ea4bdb02713a4c30aada5f3119abc85ffaa7).

SparkQA commented:
**[Test build #87512 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/87512/testReport)** for PR 20626 at commit [`10419de`](https://github.com/apache/spark/commit/10419dec3f55e16075c4a9f10df837a35225b05e).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/87512/
Test FAILed.

SparkQA commented:
**[Test build #87513 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/87513/testReport)** for PR 20626 at commit [`2d89ea4`](https://github.com/apache/spark/commit/2d89ea4bdb02713a4c30aada5f3119abc85ffaa7).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/87513/
Test PASSed.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/literals.scala:
```diff
@@ -278,40 +278,45 @@ case class Literal (value: Any, dataType: DataType) extends LeafExpression {
-      ev.isNull = "true"
-      ev.copy(s"final $javaType ${ev.value} = ${ctx.defaultValue(dataType)};")
+      val defaultValueLiteral = ctx.defaultValue(javaType) match {
+        case "null" => s"(($javaType)null)"
```
does it work for primitive types?

cloud-fan commented:
thanks, merging to master!
